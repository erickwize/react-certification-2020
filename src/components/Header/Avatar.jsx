import styled from 'styled-components';

export const Avatar = styled.img`
  background-image: url('https://upload.wikimedia.org/wikipedia/commons/b/bc/Unknown_person.jpg');
  width: 50px;
  height: 50px;
  margin-left: 15%;
  padding: 1%;
  background-size: cover;
  background-position: top center;
  border-radius: 50%;
`;
