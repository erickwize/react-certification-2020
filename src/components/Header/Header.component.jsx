import React, { useContext, useState } from 'react';
import { useHistory } from 'react-router-dom';
import HomeIcon from './img/m_logo.png';
import LoginIcon from './img/user_icon.png';
import SearchIcon from './img/search_icon.png';
import { GlobalContext } from '../../providers/GlobalContext';

// styling
import {
  Header,
  HeaderWrapper,
  HomeButton,
  Search,
  HeaderToggleWrapper,
  HeaderToggle,
  HeaderToggleLabel,
  LoginMenu,
  SearchWrapper,
  SearchSpan,
  HeaderMenuWrapper,
  HeaderMenuLink,
} from './Header.styles';

function HeaderMenu({ doSearch, dispatch }) {
  const { header, history, search, user } = useContext(GlobalContext);
  const [showMenu, setMenuVisibility] = useState(false);
  const urlHistory = useHistory();

  const updatePage = (page) => {
    const urls = {
      login: '/login',
      logout: '/',
      home: '/',
      favorites: `/user/${user.id}`,
    };
    if (page === 'logout') {
      dispatch({ type: 'REMOVE_USER_INFO', payload: {} });
    }
    setMenuVisibility(false);
    urlHistory.push(urls[page]);
  };

  const handleSearch = (ev) => {
    if (/Enter|Blur/i.test(ev.key) && search !== history) {
      doSearch(search);
      updatePage('home');
    }
  };

  const menuItems = () => {
    if (user.id) {
      return (
        <>
          <HeaderMenuLink onClick={() => updatePage('logout')}>Log out</HeaderMenuLink>
          <HeaderMenuLink onClick={() => updatePage('favorites')}>
            Favorites
          </HeaderMenuLink>
        </>
      );
    }
    return (
      <>
        <HeaderMenuLink onClick={() => updatePage('login')}>Log in</HeaderMenuLink>
      </>
    );
  };

  return (
    <Header data-testid="yt-header" background={header.background}>
      <HeaderWrapper>
        <HomeButton img={HomeIcon} onClick={() => updatePage('home')} />
        <SearchWrapper background={header.input}>
          <Search
            type="text"
            id="name"
            value={search}
            placeholder="Search..."
            onChange={({ target }) =>
              dispatch({ type: 'SET_SEARCH_KEYWORD', payload: target.value })
            }
            onBlur={() => handleSearch({ key: 'Blur' })}
            onKeyDown={handleSearch}
          />
          <SearchSpan img={SearchIcon} onClick={() => handleSearch({ key: 'Enter' })} />
        </SearchWrapper>
      </HeaderWrapper>
      <HeaderWrapper>
        <HeaderToggleWrapper>
          <HeaderToggle
            type="checkbox"
            name="darkMode"
            id="darkMode"
            onChange={({ target }) =>
              dispatch({ type: `SET_${target.checked ? 'DARK' : 'LIGHT'}_MODE` })
            }
          />
          <HeaderToggleLabel htmlFor="darkMode">Dark mode</HeaderToggleLabel>
        </HeaderToggleWrapper>
        <LoginMenu
          img={user.id ? user.avatarUrl : LoginIcon}
          onClick={() => setMenuVisibility(!showMenu)}
        />
        {showMenu ? <HeaderMenuWrapper>{menuItems()}</HeaderMenuWrapper> : null}
      </HeaderWrapper>
    </Header>
  );
}

export default HeaderMenu;
