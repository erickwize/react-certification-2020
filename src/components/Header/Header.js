import React from 'react'
import Search from '../Search/Search'
import Toggle from '../Toggle/Toggle'
import User from '../User/User'
import StyledHeader from './Header.styled'
import Hamburger from '../Hamburger/Hamburger'

const Header = React.memo(() => (
  <StyledHeader key="header">
    <div className="header__hamburger__search">
      <Hamburger />
      <Search />
    </div>
    <div className="header__theme__user">
      <Toggle />
      <User />
    </div>
  </StyledHeader>
))

export default Header
