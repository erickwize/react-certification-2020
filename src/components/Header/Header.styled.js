import styled from 'styled-components'

const StyledHeader = styled.div.attrs({
  className: 'header',
})`
  display: flex;
  background-color: #6AB1DE;
  flex-direction: row;
  justify-content:space-between;
  padding: 10px;

  box-shadow: 0.1em 0.1em 0.9em rgba(0, 0, 0, 0.5);
    .header__hamburger__search {
        display: flex;
        flex-direction: row;
    }
    .header__theme__user {
        display: flex;
        flex-direction: row;
        align-items: center;

    }
 
}
`
export default StyledHeader
