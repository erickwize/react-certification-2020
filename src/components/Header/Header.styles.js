import styled from 'styled-components';
import { VscMenu, VscSymbolBoolean } from 'react-icons/vsc';

export const Header = styled.header`
  display: flex;
  width: 100vw;
  min-height: 70px;
  padding: 17px 17px;
  background: linear-gradient(#beeaf5, #80d9df);
  margin-bottom: 1rem;
  justify-content: space-between;
  margin-bottom: 2.5rem;
`;

export const Row = styled.div`
  display: flex;
  flex-direction: row;
  width: 100vw;
  align-items: center;
`;

export const Col = styled.div`
  flex: ${(props) => props.size};
  // border: 2px solid;
  // border-color: blue;
`;

export const MenuIcon = styled(VscMenu)`
  display: block;
  font-size: 26px;
  margin-right: 20px;
  margin-left: 30px;
  color: #6c56a3;
  border: none;
  cursor: not-allowed;
`;

export const LoginContainer = styled.div`
  display: block;
  flex-direction: row;
  align-text: center;
`;
// color: #F894F1; pink

export const LogIcon = styled(VscSymbolBoolean)`
  font-size: 27px;
  margin-right: 20px;
  color: #6c56a3;
  border: none;
  transform: scaleX(-1);
  -moz-transform: scaleX(-1);
  -webkit-transform: scaleX(-1);
  -ms-transform: scaleX(-1);
  margin-left: 0.2em;
`;
