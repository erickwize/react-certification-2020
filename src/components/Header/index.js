import React, { useContext } from 'react'
import styled from 'styled-components'
import MENU_ICON from '../../assets/img/svg/menu.svg'
import USER_ICON from '../../assets/img/svg/user.svg'
import { ThemeContext } from '../../context/ThemeContext'
import SearchBox from '../SearchBox'
import SwitchTheme from '../SwitchTheme'

const HeaderStyled = styled.header`
	padding: 20px;
	display: flex;
	justify-content: space-between;
	background-color: ${props => props.theme.headerBackground};
`
const MenuButton = styled.button`
	padding: 5px 10px 0 0px;
	background: none;
	border: none;
	cursor: pointer;
	img {
		width: 70%;
	}
`
const HeaderLeft = styled.div`
	display: grid;
	grid-template-columns: 40px auto;
	width: 300px;
	@media only screen and (max-width: 768px) {
		width: 100%;
	}
`
const HeaderRight = styled.div`
	display: flex;
	@media only screen and (max-width: 768px) {
		display: none;
	}
`
const Avatar = styled.div`
	width: 30px;
	height: 30px;
	border-radius: 50%;
	overflow: hidden;
	border: 2px solid white;
	img {
		width: 100%;
	}
`

const Header = () => {
	const themeContext = useContext(ThemeContext)

	return (
		<>
			<HeaderStyled theme={themeContext}>
				<HeaderLeft>
					<MenuButton>
						<img src={MENU_ICON} alt="menu" />
					</MenuButton>
					<SearchBox />
				</HeaderLeft>
				<HeaderRight>
					<SwitchTheme />
					<Avatar>
						<img src={USER_ICON} alt="avatar" />
					</Avatar>
				</HeaderRight>
			</HeaderStyled>
		</>
	)
}

export default Header
