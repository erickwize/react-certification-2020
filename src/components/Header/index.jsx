import React from 'react';
import HeaderLogin from '../HeaderLogin';
import HeaderToggle from '../HeaderToggle';
import HeaderSearch from '../HeaderSearch';
import HeaderNav from '../HeaderNav';

const msgPlaceholder = [
  'What can we help you find?',
  'Today, you are wanna watch...',
  'If you think about it, just look it up',
];

function getRand(max) {
  return Math.round(Math.random() * max);
}

const Header = () => (
  <header className="flex flex-row bg-red-600 w-full justify-between p-2">
    <HeaderNav icon="nav-icon.png" message="Click to open menu" />
    <HeaderSearch textPlaceholder={msgPlaceholder[getRand(msgPlaceholder.length - 1)]} />
    <HeaderToggle />
    <HeaderLogin photo="login-icon.png" message="Click to login" />
  </header>
);

export default Header;
