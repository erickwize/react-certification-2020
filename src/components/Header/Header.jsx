import React, { useState, useEffect } from 'react';

import './Header.css';
import { useDispatch } from 'react-redux';
import fetchVideos from '../../actions/fetchVideos';
import changeTheme from '../../actions/themeMode';
import useModal from '../../actions/userModal';
import { useHistory } from 'react-router-dom';
import { useSelector } from 'react-redux';
import ModalComponent from '../Modal/Modal';
import Burger from '../../components/Burger/Burger';
import Menu from '../Menu/Menu';
import { useAuth } from '../../providers/Auth/Auth.provider';

const Header = () => {
  const [open, setOpen] = useState(false);
  const history = useHistory();
  const [search, setSearch] = useState("wizeline");
  const [theme, setTheme] = useState("headerMenu");
  const {isShowing, toggle} = useModal();
  const dispatch = useDispatch();
  const themeReducer = useSelector((state) => state.themeReducer);
  const menuId = "main-menu";
  const ThemeBackground = theme === 'dark' ? "white" : "black";
  const { authenticated, logout } = useAuth();
  
  useEffect(() => {
    dispatch(fetchVideos(search));
  }, []);

  const handleChange = (event) => {
    setSearch(event.target.value);
  };

  const handleTheme = () => {
    
    if(themeReducer.mode == 'light') {
      setTheme('headerMenuDark');
      dispatch(changeTheme('dark'));
    } else {
      setTheme('headerMenu');
      dispatch(changeTheme('light'));
    }
    
  }

  const handleKeyDown = (event) => {
    if (event.key === 'Enter' && search != '') {
      dispatch(fetchVideos(search));
      const path = `/`;
      history.push(path);
    }
  }

  const closeSession = () => {
    logout();
  }

  return (
    <div>
      <ModalComponent
          isShowing={isShowing}
          hide={toggle}
          theme={themeReducer.mode}
        />
    <header className={theme}>
      <div className="headerWrapper">
        <Burger open={open} setOpen={setOpen} aria-controls={menuId} theme={theme} />
        <Menu open={open} setOpen={setOpen} id={menuId} theme={theme} />
       
        <input 
          type="text" 
          id="name" 
          placeholder="Search..." 
          className="inputSearch"
          onChange={handleChange}
          onKeyPress={handleKeyDown}
          value={search}>
        </input>
      </div>
        
        <div className="headerWrapper">
          <div className="headerToggleWrapper">
            <input type="checkbox" name="darkMode" className="headerToggle" onChange={handleTheme} id="darkMode"/>
            <label htmlFor="darkMode" className="headerToggleLabel">Dark mode</label>
          </div>
          
          {authenticated ? 
            <>
            <span className="loginIcon" onClick={closeSession} />
            <div className="loginButton" onClick={closeSession}>logout</div>
            </>
            :
            <>
            <span className="loginIcon" onClick={() => toggle} />
            <div className="loginButton" onClick={toggle}>login</div>
            </>
          }
          
          
        </div>

        

      </header>

      

      </div>
    );
}

export default Header;