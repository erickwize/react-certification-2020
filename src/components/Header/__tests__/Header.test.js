import React from 'react'
import renderer from 'react-test-renderer'
import { Context } from '../../../context/context'
import Header from '../Header'

const state = {
  theme: false,
}

describe('Testing Header component', () => {
  test('renders Header without crashing', async () => {
    const rendered = renderer.create(
      <Context.Provider value={{ state }}>
        <Header />
      </Context.Provider>
    )
    expect(rendered).toMatchSnapshot()
  })
})
