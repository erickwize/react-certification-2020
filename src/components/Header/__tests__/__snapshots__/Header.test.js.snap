// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Header component renders Header without crashing 1`] = `
<div
  className="sc-eCImvq eHOHLd header"
>
  <div
    className="header__hamburger__search"
  >
    <div
      className="sc-jRQAMF NZkEi hamburger"
      onClick={[Function]}
    >
      <span />
      <span />
      <span />
    </div>
    <div
      className="sc-gsDJrp dLVeSJ"
    >
      <input
        className="sc-bdvvaa gzIPOt"
        onKeyDown={[Function]}
        role="input"
        type="text"
      />
    </div>
  </div>
  <div
    className="header__theme__user"
  >
    <div
      className="sc-dkPtyc SvsMR toggle"
    >
      <input
        checked={false}
        className="checkbox"
        id="checkbox"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="switch"
        htmlFor="checkbox"
      >
        <span
          className="slider"
        />
      </label>
      <div
        className="named"
      >
        Dark mode
      </div>
    </div>
    <div
      className="sc-hKwCoD kddDGP header__user"
    >
      <svg
        viewBox="0 0 40 40"
      >
        <path
          d="M16.207 35.5H4.5v-32h31v32H23.793L20 39.293 16.207 35.5z"
          fill="#DFF0FE"
        />
        <path
          d="M35 4v31H23.586L20 38.586 16.414 35H5V4h30zm1-1H4v33h12l4 4 4-4h12V3z"
          fill="#4788C7"
        />
        <path
          d="M24.5 15c-.052 2.448-2.051 4.406-4.5 4.406S15.552 17.448 15.5 15c.052-2.448 2.051-4.406 4.5-4.406s4.448 1.958 4.5 4.406z"
          fill="#B6DCFE"
        />
        <path
          d="M20 11c2.207 0 4 1.793 4 4s-1.793 4-4 4-4-1.793-4-4 1.793-4 4-4zm0-1c-2.761 0-5 2.239-5 5s2.239 5 5 5 5-2.239 5-5-2.239-5-5-5z"
          fill="#4788C7"
        />
        <path
          d="M12.516 28.5c.257-3.902 3.515-7 7.484-7s7.227 3.098 7.484 7H12.516z"
          fill="#98CCFD"
        />
        <path
          d="M20 22c3.52 0 6.441 2.613 6.93 6H13.07c.489-3.387 3.41-6 6.93-6zm0-1c-4.418 0-8 3.582-8 8h16c0-4.418-3.582-8-8-8z"
          fill="#4788C7"
        />
      </svg>
    </div>
  </div>
</div>
`;
